### PR TITLE
PS-5789 : Concurrent access attempts to non-shareable parallel double…

### PR DESCRIPTION
**Pull request by satya-bodapati**
…write buffer shards

PS-5789 : Concurrent access attempts to non-shareable parallel doublewrite buffer shards

Problem:
--------
With innodb_empty_free_list_algorithm=backoff, we always request page cleaners to flush
and never flush from other threads.

Buf in 8.0, startup code paths used buf_flush_sync_all_buf_pools() which did sync flushing
and this was done from multiple threads parallely and lead to concurrent access to parallel
doublewrite buffer shards.

Fix:
----
1. Like 5.7, we change buf_flush_sync_all_buf_pools() to submit request and wait for page cleaner
threads to flush instead of doing sync flushing from the threads.

2. Another special case in 8.0 is that, after page cleaners are closed, the headers are marked with
LSN (see srv_shutdown_log()). These changes have to be written and since page cleaners are already
closed, they are written/flushed via sync flushing (remember there is no other thread, that can request
flushing now). So it is safe to do this sync flushing.

startup hang issue
----------------------

When there is recovery involved and during startup we use
buf_flush_sync_all_buf_pools(), we wait for page cleaner to flush
all pages.

And page cleaners during startup, if there are tablespaces involved
in recovery, they wait for recovery flush event.

Since we do not signal on the event, the page cleaners keep waiting
and the entire startup hangs.

Fix:
---
Signal the recovery flush_start event from buf_flush_wait_end()

hang caused by SET GLOBAL innodb_page_cleaners_disabled_debug=ON (nnodb.log_flush_order test failure)
--------------------------------------------------------------------
On debug builds, When page cleaners are disabled (using SET GLOBAL
innodb_page_cleaners_disabled_debug=ON), buf_flush_sync_all_buf_pools()
infinitely waits for all pages to be flushed.

Fix:
----
When all page cleaners are disabled, fallback to flushing from
foreground threads.

### CODE REVIEW

**Comment by satya-bodapati** [comment]
https://ps80.cd.percona.com/job/percona-server-8.0-param/382/

File: storage/innobase/buf/buf0flu.cc [hunk 1]
```diff
@@ -2087,6 +2091,77 @@ bool buf_flush_do_batch(buf_pool_t *buf_pool, buf_flush_t type, ulint min_n,
...
 
+/** Waits until a flush batch of the given lsn ends
+@param[in]	new_oldest	target oldest_modified_lsn to wait for */
+static void buf_flush_wait_flushed(lsn_t new_oldest) {
+  lsn_t instance_newest[MAX_BUFFER_POOLS];
+  for (ulint i = 0; i < srv_buf_pool_instances; ++i) {
+    buf_pool_t *buf_pool;
+    instance_newest[i] = 0;
+
+    buf_pool = buf_pool_from_array(i);
```
Comment (robgolebiowski) [inline storage/innobase/buf/buf0flu.cc]: buf_pool_t *buf_pool {buf_pool_from_array(i)};

File: storage/innobase/buf/buf0flu.cc [hunk 2]
```diff
@@ -2087,6 +2091,77 @@ bool buf_flush_do_batch(buf_pool_t *buf_pool, buf_flush_t type, ulint min_n,
...
+
+    if (new_oldest == LSN_MAX) {
+      buf_flush_list_mutex_enter(buf_pool);
+
+      buf_page_t *bpage = UT_LIST_GET_FIRST(buf_pool->flush_list);
+
+      while (bpage != NULL && fsp_is_system_temporary(bpage->id.space())) {
```
Comment (robgolebiowski) [inline storage/innobase/buf/buf0flu.cc]: buf__page_t *bpage{nullptr};
for (bpage = UT_LIST_GET_FIRST(buf_pool->flush_list); bpage != nullptr && fsp_is_system_temporary(bpage->id.space(); bpage = UT_LIST_GET_NEXT(list, bpage));

File: storage/innobase/buf/buf0flu.cc [hunk 3]
```diff
@@ -2087,6 +2091,77 @@ bool buf_flush_do_batch(buf_pool_t *buf_pool, buf_flush_t type, ulint min_n,
...
+        bpage = UT_LIST_GET_NEXT(list, bpage);
+      }
+
+      if (bpage != NULL) {
```
Comment (robgolebiowski) [inline storage/innobase/buf/buf0flu.cc]: nullptr

File: storage/innobase/buf/buf0flu.cc [hunk 4]
```diff
@@ -2087,6 +2091,77 @@ bool buf_flush_do_batch(buf_pool_t *buf_pool, buf_flush_t type, ulint min_n,
...
+
+      buf_flush_list_mutex_enter(buf_pool);
+
+      buf_page_t *bpage;
+
+      /* We don't need to wait for system temporary pages */
+      for (bpage = UT_LIST_GET_LAST(buf_pool->flush_list);
+           bpage != NULL && fsp_is_system_temporary(bpage->id.space());
+           bpage = UT_LIST_GET_PREV(list, bpage)) {
+        /* Do nothing. */
```
Comment (robgolebiowski) [inline storage/innobase/buf/buf0flu.cc:2141]: you do not need this for (); is enough.

File: storage/innobase/buf/buf0flu.cc [hunk 5]
```diff
@@ -2087,6 +2091,77 @@ bool buf_flush_do_batch(buf_pool_t *buf_pool, buf_flush_t type, ulint min_n,
...
+      }
+
+      buf_flush_list_mutex_exit(buf_pool);
+    }
+  }
+
+  for (ulint i = 0; i < srv_buf_pool_instances; ++i) {
+    lsn_t oldest;
+    buf_pool_t *buf_pool;
+    buf_pool = buf_pool_from_array(i);
```
Comment (robgolebiowski) [inline storage/innobase/buf/buf0flu.cc]: buf_pool_t *buf_pool{buf_pool_from_array)i)}; or auto even.

File: storage/innobase/buf/buf0flu.cc [hunk 6]
```diff
@@ -2087,6 +2091,77 @@ bool buf_flush_do_batch(buf_pool_t *buf_pool, buf_flush_t type, ulint min_n,
...
+
+    for (;;) {
+      /* We don't need to wait for fsync of the flushed
+      blocks, because anyway we need fsync to make chekpoint.
+      So, we don't need to wait for the batch end here. */
+
+      buf_flush_list_mutex_enter(buf_pool);
+
+      buf_page_t *bpage;
```
Comment (robgolebiowski) [inline storage/innobase/buf/buf0flu.cc]: buf_page_t *bpage {nulltpr};

File: storage/innobase/buf/buf0flu.cc [hunk 7]
```diff
@@ -2087,6 +2091,77 @@ bool buf_flush_do_batch(buf_pool_t *buf_pool, buf_flush_t type, ulint min_n,
...
+
+      /* We don't need to wait for system temporary pages */
+      for (bpage = UT_LIST_GET_LAST(buf_pool->flush_list);
+           bpage != NULL && fsp_is_system_temporary(bpage->id.space());
+           bpage = UT_LIST_GET_PREV(list, bpage)) {
+        /* Do nothing. */
+      }
+
+      if (bpage != NULL) {
```
Comment (robgolebiowski) [inline storage/innobase/buf/buf0flu.cc]: oldest = bpage != nullptr ? bpage->oldest_modification : 0;
ut_ad(bpage == nullptr || bpage->in_flush_list);

File: storage/innobase/buf/buf0flu.cc (hunk 4 above)
Comment (robgolebiowski) [reply storage/innobase/buf/buf0flu.cc:2141]: nullptr

File: storage/innobase/buf/buf0flu.cc [hunk 8]
```diff
@@ -2087,6 +2091,77 @@ bool buf_flush_do_batch(buf_pool_t *buf_pool, buf_flush_t type, ulint min_n,
   return (true);
 }
 
+/** Waits until a flush batch of the given lsn ends
+@param[in]	new_oldest	target oldest_modified_lsn to wait for */
```
Comment (robgolebiowski) [inline storage/innobase/buf/buf0flu.cc]: you should write in the comment that LSN_MAX is a special value to indicate that we are flushing oldest pages from all the buffers.

File: storage/innobase/buf/buf0flu.cc [hunk 9]
```diff
@@ -2087,6 +2091,77 @@ bool buf_flush_do_batch(buf_pool_t *buf_pool, buf_flush_t type, ulint min_n,
...
+static void buf_flush_wait_flushed(lsn_t new_oldest) {
+  lsn_t instance_newest[MAX_BUFFER_POOLS];
+  for (ulint i = 0; i < srv_buf_pool_instances; ++i) {
+    buf_pool_t *buf_pool;
+    instance_newest[i] = 0;
+
+    buf_pool = buf_pool_from_array(i);
+
+    if (new_oldest == LSN_MAX) {
```
Comment (robgolebiowski) [inline storage/innobase/buf/buf0flu.cc:2106]: you can comment here also that LSN_MAX is a special marker.

**Review by robgolebiowski — CHANGES_REQUESTED** [review]

File: storage/innobase/buf/buf0flu.cc [hunk 10]
```diff
@@ -2087,6 +2091,77 @@ bool buf_flush_do_batch(buf_pool_t *buf_pool, buf_flush_t type, ulint min_n,
...
+      buf_page_t *bpage = UT_LIST_GET_FIRST(buf_pool->flush_list);
+
+      while (bpage != NULL && fsp_is_system_temporary(bpage->id.space())) {
+        bpage = UT_LIST_GET_NEXT(list, bpage);
+      }
+
+      if (bpage != NULL) {
+        ut_ad(bpage->in_flush_list);
+        instance_newest[i] = bpage->oldest_modification;
+      }
```
Comment (robgolebiowski) [inline storage/innobase/buf/buf0flu.cc]: seems that oldest_modification is always the last from the list - shouldn't you go backward then and get first page that is not fsp_is_system_temporary ?

**Comment by satya-bodapati** [comment]
> https://ps80.cd.percona.com/job/percona-server-8.0-param/382/

with recovery fix: https://ps80.cd.percona.com/job/percona-server-8.0-param/384/

**Comment by satya-bodapati** [comment]
> > https://ps80.cd.percona.com/job/percona-server-8.0-param/382/
> 
> with recovery fix: https://ps80.cd.percona.com/job/percona-server-8.0-param/384/

Failing tests (from 384):
MySQL.clone.remote_dml_compressed_row
MySQL.encryption.innodb_encryption-page-compression
MySQL.group_replication.gr_ps_applying_trx_recovery_channel
MySQL.innodb.log_flush_order
MySQL.main.backup_safe_binlog_info
MySQL.encryption.innodb_onlinealter_encryption
MySQL.auth_sec.keyring_file_data_qa
MySQL.group_replication.gr_ssl_options2
MySQL.main.bug82026
MySQL.main.mysqldumpslow
MySQL.main.percona_tls
MySQL.main.ssl_bug75311

Base trunk for comparision:
https://ps80.cd.percona.com/job/percona-server-8.0-trunk/114/#showFailuresLink


Out of these only MySQL.innodb.log_flush_order failure seems new and unique. 

This discovered a problem that when page cleaners are disabled, SET GLOBAL innodb_buf_flush_list_now=ON will hang (we wait for page cleaners to flush instead of sync flushing from that thread).

Patch to handle this is needed.

File: storage/innobase/buf/buf0flu.cc (hunk 4 above)
Comment (satya-bodapati) [reply storage/innobase/buf/buf0flu.cc:2141]: You mean for() instead of for(;;) ?

File: storage/innobase/buf/buf0flu.cc (hunk 2 above)
Comment (satya-bodapati) [reply storage/innobase/buf/buf0flu.cc]: I would like to retain the while style

File: storage/innobase/buf/buf0flu.cc (hunk 1 above)
Comment (satya-bodapati) [reply storage/innobase/buf/buf0flu.cc]: not sure this initialization is useful for pointers

Comment (satya-bodapati) [reply storage/innobase/buf/buf0flu.cc]: as I understand from the discussion, it is used to 'type narrowing' but not sure if it helps here
it allow buf_pool_t* buf_pool{0};

File: storage/innobase/buf/buf0flu.cc (hunk 4 above)
Comment (robgolebiowski) [reply storage/innobase/buf/buf0flu.cc:2141]: for (bpage = UT_LIST_GET_LAST(buf_pool->flush_list);
           bpage != nullptr && fsp_is_system_temporary(bpage->id.space());
           bpage = UT_LIST_GET_PREV(list, bpage));
You do need the { /* Do nothing. */ } part

Comment (satya-bodapati) [reply storage/innobase/buf/buf0flu.cc:2141]: got it now. thanks!

Comment (satya-bodapati) [reply storage/innobase/buf/buf0flu.cc:2141]: I see the value of this style.

by just putting ";" it looks like a mistake or unintentional error.

With explicit comment, we tell the reader, it is intentional

Would like to retain this as is

File: storage/innobase/buf/buf0flu.cc [hunk 11]
```diff
@@ -2087,6 +2091,83 @@ bool buf_flush_do_batch(buf_pool_t *buf_pool, buf_flush_t type, ulint min_n,
...
+      /* We don't need to wait for fsync of the flushed
+      blocks, because anyway we need fsync to make chekpoint.
+      So, we don't need to wait for the batch end here. */
+
+      buf_flush_list_mutex_enter(buf_pool);
+
+      buf_page_t *bpage{nullptr};
+
+      /* We don't need to wait for system temporary pages */
+      for (bpage = UT_LIST_GET_LAST(buf_pool->flush_list);
```
Comment (robgolebiowski) [inline storage/innobase/buf/buf0flu.cc:2138]: Have you considered just waiting for buf_pool->flush_list to become empty ?

File: storage/innobase/buf/buf0flu.cc (hunk 4 above)
Comment (robgolebiowski) [reply storage/innobase/buf/buf0flu.cc:2141]: as you wish

File: storage/innobase/buf/buf0flu.cc (hunk 11 above)
Comment (satya-bodapati) [reply storage/innobase/buf/buf0flu.cc:2138]: Yes, it seems we are trying to optimize by not waiting for temp tablespace pages to be flushed.

So, by doing that (just waiting for flush_list to become empty), we will remove optimization and make it worse(may be)

Comment (robgolebiowski) [reply storage/innobase/buf/buf0flu.cc:2138]: yes, makes sense.

**Review by robgolebiowski — APPROVED** [review]

**Comment by satya-bodapati** [comment]
I have analyzed travis failures. They are not caused by this commit
1. protobuf failure because of old version on system
2. clone0api.cc:1839:36: error: suggest braces around initialization of subobject [-Werror,-Wmissing-braces] happens with clang4 and clang 5 but not with higher versions. happens on PS as well
3. void* __builtin_memcpy(void*, const void*, long unsigned int)’ specified bound 18446744073709551615 exceeds maximum object size 9223372036854775807 [-Werror=stringop-overflow=] : happens with clang 9. Known issue. Przemek will push fix for this

**Pull request merged**